sdb/commands/stacks.py

```diff
@@ -0,0 +1,255 @@
+    # by module name, we find the segment in memory where this module
+    # resides and do the matching based on the address of the function
+    # of the current frame.
+    #
+    @staticmethod
+    def find_module_memory_segment(prog: drgn.Program,
+                                   mod_name: str) -> Tuple[int, int]:
+        """
+        Looks for the segment in memory where `mod_name` is
+        loaded within `prog`.
+
+        Returns:
+            (<base_offset>, <size>) if `mod_name` is found.
+            (-1, 0) otherwise.
+        """
+        for mod in list_for_each_entry('struct module',
+                                       prog['modules'].address_of_(), 'list'):
+            if mod.name.string_().decode("utf-8") == mod_name:
+                return (mod.core_layout.base.value_(),
+                        mod.core_layout.size.value_())
+        return (-1, 0)
+
+    def validate_args(self, prog: drgn.Program,
+                      args: argparse.Namespace) -> None:
+        if args.function:
+            try:
+                func = prog[args.function]
+            except KeyError:
+                raise sdb.CommandError(
+                    self.name,
+                    "symbol '{:s}' does not exist".format(args.function))
+            if func.type_.kind != drgn.TypeKind.FUNCTION:
+                raise sdb.CommandError(
+                    self.name, "'{:s}' is not a function".format(args.function))
+
+        task_states = Stacks.TASK_STATES.values()
+        if args.tstate and not args.tstate in task_states:
+            raise sdb.CommandError(
+                self.name,
+                "'{:s}' is not a valid task state (acceptable states: {:s})".
+                format(args.tstate, ", ".join(task_states)))
+
+        if args.module and Stacks.find_module_memory_segment(
+                self.prog, args.module)[0] == -1:
+            raise sdb.CommandError(
+                self.name,
+                "module '{:s}' doesn't exist or isn't currently loaded".format(
+                    args.module))
+
+    def call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
+        # pylint: disable=too-many-locals
+        # pylint: disable=too-many-branches
+
+        #
+        # As the exception explains the code that follows this statement only
+        # works for linux kernel targets (crash dumps or live systems). When
+        # support for userland is added we can factor the kernel code that
+        # follows into its own function and switch to the correct codepath
+        # depending on the target.
+        #
+        if not self.prog.flags & drgn.ProgramFlags.IS_LINUX_KERNEL:
+            raise sdb.CommandError(self.name,
+                                   "userland targets are not supported yet")
+        self.validate_args(self.prog, self.args)
+
+        mod_start, mod_end = -1, -1
+        if self.args.module:
+            mod_start, mod_size = Stacks.find_module_memory_segment(
```

Comment:
I don't know if it really matters (at least right now), but should this be moved into the constructor for this class?

I guess this could be asked for other lines in this function too.. IMO, it's just a little awkward to me how we have "initialization logic" in this `call` method.. but I think that's partially a result of how `call` isn't structured to iterate over the input object, but rather it's a single function that "does everything" (sets up initial state, finds all tasks, prints them, etc.).

I'm OK with this; the structure is just a little bit different than other command classes (where initial state would be set up in the constructor, etc).